File: 2025-03-10/Binary_Tree_Level_Order_Traversal.java
import java.util.*;

public class Binary_Tree_Level_Order_Traversal {
    public static List<List<Integer>> levelOrder(TreeNode root) {
        if(root == null) return new ArrayList<>();

        List<List<Integer>> ans = new ArrayList<>();
        Queue<TreeNode> q = new LinkedList<>();
        q.offer(root);

        while(!q.isEmpty()) {
            int size = q.size();
            List<Integer> cur = new ArrayList<>();

            for(int i=0;i<size;i++) {
                TreeNode node = q.poll();
                cur.add(node.val);

                if(node.left != null) q.offer(node.left);
                if(node.right != null) q.offer(node.right);
            }
            ans.add(cur);
        }
        return ans;
    }
    public static void main(String[] args) {
        Integer[] arr = new Integer[]{3,9,20,null,null,15,7};
        Tree tree = new Tree();
        TreeNode root = tree.arrayToTree(arr);
        List<List<Integer>> ans = levelOrder(root);
        System.out.println(ans.toString());
    }
}

File: 2025-09-29/Binary_Tree_Level_Order_Traversal.py
from collections import deque
from typing import Optional, List

from TreeNode import TreeNode
from Tree import Tree


class Solution:
    def levelOrder(self, root: Optional[TreeNode]) -> List[List[int]]:
        level = []

        q = deque()
        q.append(root)

        while q:
            curLevel = []
            for i in range(len(q)):
                node = q.popleft()
                curLevel.append(node.val)

                if node.left:
                    q.append(node.left)

                if node.right:
                    q.append(node.right)

            level.append(curLevel)

        return level



sol = Solution()
tree = Tree()
node = tree.array_to_tree([3,9,20,None,None,15,7])
print(sol.levelOrder(node))





# [[],[],[]]